### PR TITLE
Make elaboration generic over input

### PR DESCRIPTION
**Pull request by compiler-errors**
Combines all the `elaborate_*` family of functions into just one, which is an iterator over the same type that you pass in (e.g. elaborating `Predicate` gives `Predicate`s, elaborating `Obligation`s gives `Obligation`s, etc.)

### CODE REVIEW

File: compiler/rustc_hir_analysis/src/check/compare_impl_item.rs [hunk 1]
```diff
@@ -2034,7 +2034,7 @@ pub(super) fn check_type_bounds<'tcx>(
         ObligationCause::new(impl_ty_span, impl_ty_def_id, code)
     };
 
-    let obligations = tcx
+    let obligations: Vec<_> = tcx
```
Comment (b-naber) [inline compiler/rustc_hir_analysis/src/check/compare_impl_item.rs:2037]: We don't have to `collect` here anymore, do we?

**Review by b-naber — APPROVED** [review]
Looks nice. One small nit then r=me.

File: compiler/rustc_hir_analysis/src/check/compare_impl_item.rs (hunk 1 above)
Comment (compiler-errors) [reply compiler/rustc_hir_analysis/src/check/compare_impl_item.rs:2037]: We don't *need* to, but there's a debug line right below.

Comment (compiler-errors) [reply compiler/rustc_hir_analysis/src/check/compare_impl_item.rs:2037]: `compare_impl_item` isn't super perf sensitive anyways.

**Comment by nnethercote** [comment]
Not enough perf changes here to be a problem.

@rustbot label: +perf-regression-triaged

**Pull request merged**